### PR TITLE
Fix for getting values of pointers in traces in new SMT backend

### PR DESCRIPTION
**Pull request by NlightNFotis**
This PR changes the code that constructs bit vector terms from solver
values so that it now supports pointers.

This allows us to now see pointer values in traces in the new SMT backend.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/cbmc-incr-smt2/pointers/pointer_values.desc [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+pointer_values.c
+--trace
+\[main\.assertion\.1\] line \d should fail because a can be any pointer val, including NULL: FAILURE
+\[main\.assertion\.2\] line \d should fail because a and c can point to same object: FAILURE
+a=\(\(signed int \*\)NULL\)
+c=\(signed int \*\)1
+^EXIT=10$
+^SIGNAL=0$
+--
+--
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/pointers/pointer_values.desc:11]: :pick: Missing test description - "Test printing of NULL pointer values in trace."

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp [hunk 2]
```diff
@@ -44,7 +44,7 @@ class value_expr_from_smt_factoryt : public smt_term_const_downcast_visitort
     {
       INVARIANT(
         pointer_type->get_width() == sort_width,
-        "Width of smt bit vector term must match the width of pointer  type.");
+        "Width of smt bit vector term must match the width of pointer type.");
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:47]: :pick: This change appears to have been squashed into the wrong commit.

File: unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp [hunk 3]
```diff
@@ -68,6 +68,16 @@ TEST_CASE("Value expr construction from smt.", "[core][smt2_incremental]")
...
+    rowt{
+      smt_bit_vector_constant_termt{0, 64},
+      from_integer(0, pointer_typet(unsignedbv_typet(64), 64))},
+    // The reason for the more intricate elaboration of a pointer with a value
+    // of 12 is a limitation in the design of from_integer, which only handles
+    // pointers with value 0 (null pointers).
+    rowt{
+      smt_bit_vector_constant_termt{12, 64},
+      constant_exprt(
+        integer2bvrep(12, 64), pointer_typet(unsignedbv_typet(64), 64))},
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: :bulb: It could be a bit tidier if these pointer tests were separated out into a macro like for the BIT_VECTOR_TESTS.
:bulb: These tests could be made more thorough by re-running them for 32 and 64 bit pointer widths.
:bulb: Using a pointer to a type such as `bool_typet` or `void_type()` instead of a pointer to `unsignedbv_typet(64)`, may help improve the readability of the test by making all instances of the number `64` relate to the width of the pointer, rather than the width of the type the pointer points to.

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp [hunk 4]
```diff
@@ -37,13 +37,35 @@ class value_expr_from_smt_factoryt : public smt_term_const_downcast_visitort
...
+      INVARIANT(
+        pointer_type->get_width() == sort_width,
+        "Width of smt bit vector term must match the width of pointer type.");
+      if(bit_vector_constant.value() == 0)
+      {
+        result = null_pointer_exprt{*pointer_type};
+      }
+      else
+      {
+        // The reason we are mnually constructing a constant_exprt here is a
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: :pick: Typo "mnually" should be "manually"

**Review by thomasspriggs — APPROVED** [review]

**Review by TGWDB — APPROVED** [review]
